DataManagementSystem/DB/FileCatalogComponents/DirectoryTreeBase.py

```diff
@@ -336,148 +337,156 @@ def getDirectoryParameters( self, path ):
+                                           self.db.fileManager.setFileGroup,
+                                           recursive = recursive )
 
 #####################################################################
-  def setDirectoryMode( self, path, mode ):
-    """ set the directory mask
+  def _setDirectoryMode( self, path, mode ):
+    """ set the directory mode
+
+        :param mixed path: directory path as a string or int or list of ints or select statement
+        :param int mode: new mode
     """
     return self._setDirectoryParameter( path, 'Mode', mode )
 
 #####################################################################
-  def changeDirectoryMode( self, paths, s_uid = 0, s_gid = 0 ):
-    """ Bulk setting of the directory owner
+  def changeDirectoryMode( self, paths, recursive = False ):
+    """ Bulk setting of the directory mode
+
+        :param dictionary paths : dictionary < lfn : mode >
     """
-    result = self.db.ugManager.findUser( s_uid )
-    if not result['OK']:
-      return result
-    uid = result['Value']
-    result = self.db.ugManager.findGroup( s_gid )
-    if not result['OK']:
-      return result
-    gid = result['Value']
+    return self._changeDirectoryParameter( paths,
+                                           self._setDirectoryMode,
+                                           self.db.fileManager.setFileMode,
+                                           recursive = recursive )
 
-    result = checkArgumentFormat( paths )
-    if not result['OK']:
-      return result
-    arguments = result['Value']
+#####################################################################
+  def _changeDirectoryParameter( self, paths,
+                                 directoryFunction,
+                                 fileFunction,
+                                 recursive = False ):
+    """ Bulk setting of the directory parameter with recursion for all the subdirectories and files
+
+        :param dictionary paths : dictionary < lfn : value >, where value is the value of parameter to be set
+        :param function directoryFunction: function to change directory(ies) parameter
+        :param function fileFunction: function to change file(s) parameter
+        :param bool recursive: flag to apply the operation recursively
+    """
+    arguments = paths
     successful = {}
     failed = {}
-    for path, mode in arguments.items():
-      result = self.setDirectoryMode( path, mode )
+    for path, attribute in arguments.items():
+      result = directoryFunction( path, attribute )
       if not result['OK']:
         failed[path] = result['Message']
-      else:
-        successful[path] = True
+        continue
+      if recursive:
+        result = self.__getDirID( path )
+        if not result['OK']:
+          failed[path] = result['Message']
+          continue
+        dirID = result['Value']
+        result = self.getSubdirectoriesByID( dirID, requestString = True, includeParent = True )
```

Comment:
In line 453 the call is just to change the directory attribute which is OK if no recursive flag. Here the result is used also for contained files including the parent directory.